qiskit_nature/second_q/algorithms/initial_points/mp2_initial_point.py

```diff
@@ -14,75 +14,101 @@
+
+    """
+    # We use NumPy broadcasting to compute the matrix of occupied - virtual energy deltas with
+    # shape (num_occ, num_vir), such that
+    # energy_deltas[i, a] = orbital_energy[i] - orbital_energy[a].
+    # NOTE In the unrestricted-spin calculation, the orbital energies will be a 2D array, and this
+    # logic will need to be revisited.
+    energy_deltas = orbital_energies[:num_occ, np.newaxis] - orbital_energies[num_occ:]
+
+    # We now want to compute a 4D tensor of (occupied, occupied) - (virtual, virtual)
+    # energy deltas with shape (num_occ, num_vir, num_occ, num_vir), such that
+    # double_deltas[i, a, j, b] = orbital_energies[i] + orbital_energies[j]
+    #                             - orbital_energies[a] - orbital_energies[b].
+    # Again we can use NumPy broadcasting to speed this up.
+    double_deltas = energy_deltas[:, :, np.newaxis, np.newaxis] + energy_deltas
+
+    # Create integral matrix that uses occupied and virtual indices rather than MO indices.
+    integral_matrix_ovov = integral_matrix[:num_occ, num_occ:, :num_occ, num_occ:]
+
+    # Compute T2 amplitudes and transpose to num_occ, num_occ, num_vir, num_vir.
+    t2_amplitudes = (integral_matrix_ovov / double_deltas).transpose(0, 2, 1, 3)
+
+    # Compute MP2 energy correction.
+    energy_correction = np.einsum("ijab,iajb", t2_amplitudes, integral_matrix_ovov) * 2
+    energy_correction -= np.einsum("ijab,ibja", t2_amplitudes, integral_matrix_ovov)
+
+    return t2_amplitudes, energy_correction
 
 
 class MP2InitialPoint(InitialPoint):
     """Compute the second-order Møller-Plesset perturbation theory (MP2) initial point.
 
     The computed MP2 correction coefficients are intended for use as an initial point for the VQE
-    parameters in combination with a :class:`~qiskit_nature.second_q.circuit.library.ansatzes.ucc.UCC`
-    ansatz.
+    parameters in combination with a
+    :class:`~qiskit_nature.second_q.circuit.library.ansatzes.ucc.UCC` ansatz.
 
-    The coefficients and energy corrections are computed using the :meth:`compute` method, which
-    requires the :attr:`grouped_property` and :attr:`ansatz` to be passed as arguments or the
+    The initial point parameters are computed using the :meth:`compute` method, which requires the
+    :attr:`grouped_property` and :attr:`ansatz` to be passed as arguments or the
     :attr:`grouped_property` and :attr:`excitation_list` attributes to be set already.
 
-    ``MP2InitialPoint`` requires the
-    :class:`~qiskit_nature.second_q.properties.ElectronicEnergy`, which should
-    be passed in via the :attr:`grouped_property` attribute. From this it must obtain the two-body
-    molecular orbital electronic integrals and orbital energies. If the Hartree-Fock reference
-    energy is also obtained, it will be used to compute the absolute MP2 energy using the
-    :meth:`get_energy` method.
+    ``MP2InitialPoint`` requires the :class:`~qiskit_nature.second_q.properties.ElectronicEnergy`,
```

Comment:
This should be updated to also reflect that the `ParticleNumber` is also being used :+1: 

Note: with the upcoming refactoring, this will change slightly since the number of particles will actually be stored on the problem itself. So for now you can simply use this as done now, but this will get improved a bit more shortly :+1: